### PR TITLE
fix: scroll restoration 

### PR DESCRIPTION
**Pull request by AkifumiSato**
When `experimental.scrollRestoration = true`, the restoration position may not be appropriate on the screen when `getServerSidePropos` is executed.
This is due to the fact that the `_idx` inside the Router is sequentially numbered.

This modification restores the position of the browser back after reloading.

https://user-images.githubusercontent.com/25711332/161005840-821f4811-067f-4cd3-8355-abfe8f065574.mov

https://user-images.githubusercontent.com/25711332/161005887-9a2fdaef-b3ca-460a-9e03-fee45e7a8ea2.mov

Also, the key of the history is published from `Router`. (see #34980)

## Bug

- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Errors have helpful link attached, see `contributing.md`

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [ ] Related issues linked using `fixes #number`
- [x] Integration tests added
- [x] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have helpful link attached, see `contributing.md`

## Documentation / Examples

- [ ] Make sure the linting passes by running `yarn lint`

### CODE REVIEW

**Review by ijjk — COMMENTED** [review]
Hi, can we add a test case for this?

**Comment by AkifumiSato** [comment]
@ijjk Added testing.

File: packages/next/shared/lib/router/router.ts [hunk 1]
```diff
@@ -804,29 +811,29 @@ export default class Router implements BaseRouter {
...
-            const v = sessionStorage.getItem('__next_scroll_' + idx)
+            const v = sessionStorage.getItem('__next_scroll_' + key)
             forcedScroll = JSON.parse(v!)
           } catch {
             forcedScroll = { x: 0, y: 0 }
           }
         }
       }
     }
-    this._idx = idx
```
Comment (ijjk) [inline packages/next/shared/lib/router/router.ts:829]: Why was this changed from index based to a random key?

Comment (AkifumiSato) [reply packages/next/shared/lib/router/router.ts:829]: With index, it is reset to 0 when reloading, which may cause inconsistencies.

Comment (AkifumiSato) [reply packages/next/shared/lib/router/router.ts:829]: 1. Load `/person/1` (SSR)
```
+-----------+
| /person/1 |
| idx: 0    |
+-----------+
```
2. Move to `/person/2` (pushState)
```
+-----------+-----------+
| /person/1 | /person/2 |
| idx: 0    | idx: 1    |
+-----------+-----------+
```
3. Reload `/person/2` (SSR)
```
+-----------+-----------+
| /person/1 | /person/2 |
| idx: 0    | idx: 0    |
+-----------+-----------+
```
The `idx` value and scroll position are accidentally shared between pages.

**Comment by AkifumiSato** [comment]
@ijjk Are there any other code modifications needed?

File: docs/api-reference/next/router.md [hunk 2]
```diff
@@ -52,6 +52,7 @@ The following is the definition of the `router` object returned by both [`useRou
 - `domainLocales`: `Array<{domain, defaultLocale, locales}>` - Any configured domain locales.
 - `isReady`: `boolean` - Whether the router fields are updated client-side and ready for use. Should only be used inside of `useEffect` methods and not for conditionally rendering on the server. See related docs for use case with [automatically statically optimized pages](/docs/advanced-features/automatic-static-optimization.md)
 - `isPreview`: `boolean` - Whether the application is currently in [preview mode](/docs/advanced-features/preview-mode.md).
+- `key`: `string` - String uniquely representing the current history entry. Reloading will reset it.
```
Comment (ijjk) [inline docs/api-reference/next/router.md:55]: I don't think we should expose this value publicly as it could cause users to rely on it in unintended ways.

File: packages/next/server/render.tsx [hunk 3]
```diff
@@ -149,6 +151,7 @@ class ServerRouter implements NextRouter {
     this.domainLocales = domainLocales
     this.isPreview = !!isPreview
     this.isLocaleDomain = !!isLocaleDomain
+    this.key = createKey()
```
Comment (ijjk) [inline packages/next/server/render.tsx:154]: We shouldn't need to have a unique key during SSR as the key is meant to be used on the client only right?

File: docs/api-reference/next/router.md (hunk 2 above)
Comment (AkifumiSato) [reply docs/api-reference/next/router.md:55]: It is suggested to publish the history key in #34980 .
I also need this feature because I want to associate state and history and store them in storage.

File: packages/next/server/render.tsx (hunk 3 above)
Comment (AkifumiSato) [reply packages/next/server/render.tsx:154]: Right.
Should it be `null` in SSR?

File: docs/api-reference/next/router.md (hunk 2 above)
Comment (ijjk) [reply docs/api-reference/next/router.md:55]: The mentioned use case in the discussion wouldn't be solved here since the key is random and wouldn't give a signal to how many navigations or where in the navigation history the user is.

File: packages/next/server/render.tsx (hunk 3 above)
Comment (ijjk) [reply packages/next/server/render.tsx:154]: I think leaving this unset during SSR would be fine and doesn't need to be set anyways if it's not exposed publicly.

File: docs/api-reference/next/router.md (hunk 2 above)
Comment (AkifumiSato) [reply docs/api-reference/next/router.md:55]: What about in terms of providing functionality equivalent to remix and WICG's navigation-api? This feature would solve the case where you want to link state and history, etc.

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | AkifumiSato/next.js history-unique-key | Change |
| - | - | - | - |
| buildDuration | 15.2s | 15.6s | ⚠️ +459ms |
| buildDurationCached | 6.3s | 6.2s | -112ms |
| nodeModulesSize | 484 MB | 484 MB | ⚠️ +526 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | AkifumiSato/next.js history-unique-key | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.212 | 3.175 | -0.04 |
| / avg req/sec | 778.29 | 787.5 | +9.21 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.357 | 1.296 | -0.06 |
| /error-in-render avg req/sec | 1841.86 | 1929.2 | +87.34 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | AkifumiSato/next.js history-unique-key | Change |
| - | - | - | - |
| 925.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42 kB | 42 kB | ✓ |
| main-HASH.js gzip | 28.3 kB | 28.4 kB | ⚠️ +34 B |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 72 kB | 72 kB | ⚠️ +34 B |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | AkifumiSato/next.js history-unique-key | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | AkifumiSato/next.js history-unique-key | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.36 kB | 1.36 kB | ✓ |
| _error-HASH.js gzip | 192 B | 192 B | ✓ |
| amp-HASH.js gzip | 309 B | 309 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 3.04 kB | 3.04 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 920 B | 920 B | ✓ |
| image-HASH.js gzip | 5.74 kB | 5.74 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 2.36 kB | 2.36 kB | ✓ |
| routerDirect..HASH.js gzip | 320 B | 320 B | ✓ |
| script-HASH.js gzip | 392 B | 392 B | ✓ |
| withRouter-HASH.js gzip | 319 B | 319 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16 kB | 16 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | AkifumiSato/next.js history-unique-key | Change |
| - | - | - | - |
| _buildManifest.js gzip | 461 B | 461 B | ✓ |
| Overall change | 461 B | 461 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | AkifumiSato/next.js history-unique-key | Change |
| - | - | - | - |
| index.html gzip | 530 B | 532 B | ⚠️ +2 B |
| link.html gzip | 544 B | 545 B | ⚠️ +1 B |
| withRouter.html gzip | 524 B | 526 B | ⚠️ +2 B |
| Overall change | 1.6 kB | 1.6 kB | ⚠️ +5 B |

</details>

#### Diffs
<details>
<summary>Diff for <strong>main-HASH.js</strong></summary>

```diff
@@ -2675,6 +2675,7 @@
         "locales",
         "defaultLocale",
         "isReady",
+        "key",
         "isPreview",
         "isLocaleDomain",
         "domainLocales"
@@ -4394,6 +4395,7 @@
       exports.isLocalURL = isLocalURL;
       exports.interpolateAs = interpolateAs;
       exports.resolveHref = resolveHref;
+      exports.createKey = createKey;
       exports["default"] = void 0;
       var _normalizeTrailingSlash = __webpack_require__(2700);
       var _routeLoader = __webpack_require__(2497);
@@ -4789,6 +4791,11 @@
             throw err;
           }));
       }
+      function createKey() {
+        return Math.random()
+          .toString(36)
+          .slice(2, 10);
+      }
       var Router = /*#__PURE__*/ (function() {
         function Router(pathname1, query1, as1, param) {
           var initialProps = param.initialProps,
@@ -4813,7 +4820,6 @@
           this.sdr = {};
           // In-flight middleware preflight requests
           this.sde = {};
-          this._idx = 0;
           this.onPopState = function(e) {
             var state = e.state;
             if (!state) {
@@ -4845,11 +4851,11 @@
             var url = state.url,
               as = state.as,
               options = state.options,
-              idx = state.idx;
+              key = state.key;
             if (false) {
               var v;
             }
-            _this._idx = idx;
+            _this.key = key;
             var pathname2 = (0, _parseRelativeUrl).parseRelativeUrl(url)
               .pathname;
             // Make sure we don't re-render on initial load,
@@ -4924,6 +4930,7 @@
             (self.__NEXT_DATA__.appGip && !self.__NEXT_DATA__.gsp) ||
             (!autoExportDynamic && !self.location.search && !false)
           );
+          this.key = createKey();
           if (false) {
           }
           this.state = {
@@ -5762,8 +5769,8 @@
                     as: as,
                     options: options,
                     __N: true,
-                    idx: (this._idx =
-                      method !== "pushState" ? this._idx : this._idx + 1)
+                    key: (this.key =
+                      method !== "pushState" ? this.key : createKey())
                   }, // Passing the empty string here should be safe against future changes to the method.
                   // https://developer.mozilla.org/en-US/docs/Web/API/History/replaceState
                   "",
```
</details>
<details>
<summary>Diff for <strong>index.html</strong></summary>

```diff
@@ -19,7 +19,7 @@
       defer=""
     ></script>
     <script
-      src="/_next/static/chunks/main-afb4b7219fe8d317.js"
+      src="/_next/static/chunks/main-e52d02ce54a2a698.js"
       defer=""
     ></script>
     <script
```
</details>
<details>
<summary>Diff for <strong>link.html</strong></summary>

```diff
@@ -19,7 +19,7 @@
       defer=""
     ></script>
     <script
-      src="/_next/static/chunks/main-afb4b7219fe8d317.js"
+      src="/_next/static/chunks/main-e52d02ce54a2a698.js"
       defer=""
     ></script>
     <script
```
</details>
<details>
<summary>Diff for <strong>withRouter.html</strong></summary>

```diff
@@ -19,7 +19,7 @@
       defer=""
     ></script>
     <script
-      src="/_next/static/chunks/main-afb4b7219fe8d317.js"
+      src="/_next/static/chunks/main-e52d02ce54a2a698.js"
       defer=""
     ></script>
     <script
```
</details>
</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | AkifumiSato/next.js history-unique-key | Change |
| - | - | - | - |
| buildDuration | 18.1s | 18.6s | ⚠️ +497ms |
| buildDurationCached | 6s | 6s | ⚠️ +59ms |
| nodeModulesSize | 484 MB | 484 MB | ⚠️ +526 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | AkifumiSato/next.js history-unique-key | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.091 | 3.194 | ⚠️ +0.1 |
| / avg req/sec | 808.82 | 782.75 | ⚠️ -26.07 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.352 | 1.278 | -0.07 |
| /error-in-render avg req/sec | 1848.68 | 1956.1 | +107.42 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | AkifumiSato/next.js history-unique-key | Change |
| - | - | - | - |
| 925.HASH.js gzip | 178 B | 178 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.7 kB | 28.7 kB | ⚠️ +35 B |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.6 kB | 72.7 kB | ⚠️ +35 B |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | AkifumiSato/next.js history-unique-key | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | AkifumiSato/next.js history-unique-key | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 179 B | 179 B | ✓ |
| amp-HASH.js gzip | 313 B | 313 B | ✓ |
| css-HASH.js gzip | 325 B | 325 B | ✓ |
| dynamic-HASH.js gzip | 3.02 kB | 3.02 kB | ✓ |
| head-HASH.js gzip | 351 B | 351 B | ✓ |
| hooks-HASH.js gzip | 921 B | 921 B | ✓ |
| image-HASH.js gzip | 5.78 kB | 5.78 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 2.44 kB | 2.44 kB | ✓ |
| routerDirect..HASH.js gzip | 322 B | 322 B | ✓ |
| script-HASH.js gzip | 393 B | 393 B | ✓ |
| withRouter-HASH.js gzip | 317 B | 317 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 16.1 kB | 16.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | AkifumiSato/next.js history-unique-key | Change |
| - | - | - | - |
| _buildManifest.js gzip | 459 B | 459 B | ✓ |
| Overall change | 459 B | 459 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | AkifumiSato/next.js history-unique-key | Change |
| - | - | - | - |
| index.html gzip | 533 B | 533 B | ✓ |
| link.html gzip | 547 B | 546 B | -1 B |
| withRouter.html gzip | 527 B | 526 B | -1 B |
| Overall change | 1.61 kB | 1.6 kB | -2 B |

</details>

#### Diffs
<details>
<summary>Diff for <strong>main-HASH.js</strong></summary>

```diff
@@ -2675,6 +2675,7 @@
         "locales",
         "defaultLocale",
         "isReady",
+        "key",
         "isPreview",
         "isLocaleDomain",
         "domainLocales"
@@ -4394,6 +4395,7 @@
       exports.isLocalURL = isLocalURL;
       exports.interpolateAs = interpolateAs;
       exports.resolveHref = resolveHref;
+      exports.createKey = createKey;
       exports["default"] = void 0;
       var _normalizeTrailingSlash = __webpack_require__(2700);
       var _routeLoader = __webpack_require__(2497);
@@ -4789,6 +4791,11 @@
             throw err;
           }));
       }
+      function createKey() {
+        return Math.random()
+          .toString(36)
+          .slice(2, 10);
+      }
       var Router = /*#__PURE__*/ (function() {
         function Router(pathname1, query1, as1, param) {
           var initialProps = param.initialProps,
@@ -4813,7 +4820,6 @@
           this.sdr = {};
           // In-flight middleware preflight requests
           this.sde = {};
-          this._idx = 0;
           this.onPopState = function(e) {
             var state = e.state;
             if (!state) {
@@ -4845,11 +4851,11 @@
             var url = state.url,
               as = state.as,
               options = state.options,
-              idx = state.idx;
+              key = state.key;
             if (false) {
               var v;
             }
-            _this._idx = idx;
+            _this.key = key;
             var pathname2 = (0, _parseRelativeUrl).parseRelativeUrl(url)
               .pathname;
             // Make sure we don't re-render on initial load,
@@ -4924,6 +4930,7 @@
             (self.__NEXT_DATA__.appGip && !self.__NEXT_DATA__.gsp) ||
             (!autoExportDynamic && !self.location.search && !false)
           );
+          this.key = createKey();
           if (false) {
           }
           this.state = {
@@ -5762,8 +5769,8 @@
                     as: as,
                     options: options,
                     __N: true,
-                    idx: (this._idx =
-                      method !== "pushState" ? this._idx : this._idx + 1)
+                    key: (this.key =
+                      method !== "pushState" ? this.key : createKey())
                   }, // Passing the empty string here should be safe against future changes to the method.
                   // https://developer.mozilla.org/en-US/docs/Web/API/History/replaceState
                   "",
```
</details>
<details>
<summary>Diff for <strong>index.html</strong></summary>

```diff
@@ -19,7 +19,7 @@
       defer=""
     ></script>
     <script
-      src="/_next/static/chunks/main-afb4b7219fe8d317.js"
+      src="/_next/static/chunks/main-e52d02ce54a2a698.js"
       defer=""
     ></script>
     <script
```
</details>
<details>
<summary>Diff for <strong>link.html</strong></summary>

```diff
@@ -19,7 +19,7 @@
       defer=""
     ></script>
     <script
-      src="/_next/static/chunks/main-afb4b7219fe8d317.js"
+      src="/_next/static/chunks/main-e52d02ce54a2a698.js"
       defer=""
     ></script>
     <script
```
</details>
<details>
<summary>Diff for <strong>withRouter.html</strong></summary>

```diff
@@ -19,7 +19,7 @@
       defer=""
     ></script>
     <script
-      src="/_next/static/chunks/main-afb4b7219fe8d317.js"
+      src="/_next/static/chunks/main-e52d02ce54a2a698.js"
       defer=""
     ></script>
     <script
```
</details>
</details>
Commit: d0a70e360090a494c41d571ad3d6898cc9614263

**Comment by AkifumiSato** [comment]
The test cases that were added in the local environment were also flaky test and have been fixed.

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 993e2d2a6aa9abb16ddb4e9771def86b65cb312b

`yarn testheadless test/integration/cli/test/index.test.js` 
- CLI Usage > info > should print output

<details>
<summary>Expand output</summary>

  ● CLI Usage › info › should print output

    expect(received).toBe(expected) // Object.is equality

    Expected: ""
    Received: "warn  - Latest canary version not detected, detected: \"12.1.6-canary.2\", newest: \"12.1.6-canary.4\".
            Please try the latest canary version (`npm install next@canary`) to confirm the issue still exists before creating a new issue.
            Read more - https://nextjs.org/docs/messages/opening-an-issue
    "

      496 |       // warning will show so skip this check for the stable release
      497 |       if (pkg.version.includes('-canary')) {
    > 498 |         expect(info.stderr || '').toBe('')
          |                                   ^
      499 |       }
      500 |       expect(info.stdout).toMatch(
      501 |         new RegExp(`

      at Object.<anonymous> (integration/cli/test/index.test.js:498:35)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by AkifumiSato** [comment]
@ijjk The error in the test was not reproduced in my environment. I thought that it was probably a case of the system failing if the restore was not completed in time, so I fixed it to wait until `routeChangeComplete`.
If this still falls through, please advise.

**Comment by AkifumiSato** [comment]
I have created other PR that has been modified to not public the `key` and will close it.
https://github.com/vercel/next.js/pull/36861
